# 遇到过的问题

> 执行`spark-submit wordcount.py`，出现了`py4j.protocol.Py4JJavaError`错误。

- 没启动hadoop
- 文件目录设置错了。待读取的文件存放在本地，但这样设置`/in/wc.txt`。因为spark默认读的是hdfs中的数据。

---------------------------------------------------------------------------------

> 本地执行spark `WordCount.java`，报错`NoSuchMethodError: io.netty.buffer.PooledByteBufAllocator.defaultNumHeapArena`

io.netty 的 两个 jar 包导致的冲突，所以剔除 spark-core 中所有的低版本的 netty，再添加依赖高版本的：

```xml
<dependency>
    <groupId>org.apache.spark</groupId>
    <artifactId>spark-core_2.11</artifactId>
    <version>2.4.4</version>
    <!--将netty包排除-->
    <exclusions>
        <exclusion>
        	<groupId>io.netty</groupId>
            <artifactId>netty</artifactId>
        </exclusion>
    </exclusions>
    </dependency>
    <dependency>
        <groupId>io.netty</groupId>
        <artifactId>netty-all</artifactId>
        <version>4.1.42.Final</version>
    </dependency>
```

---------------------------------------------------------------------------------

> 提交集群执行`spark-submit --master spark://zgg:7077 --class sparkbase.wordcount hdfs://zgg:9000/in/wc.txt`，报错`java.lang.ClassNotFoundException: sparkbase.wordcount`

解决方法见`提交集群运行过程.md`文档

---------------------------------------------------------------------------------

> 启动mysql服务，失败，出现`Job for mysqld.service failed because the control process exited with error code.`

var/lib/mysql 这个目标路径已经存在，导致无法初始化，删除即可。

```sh
[root@zgg script]# rm -rf /var/lib/mysql
[root@zgg script]# /bin/systemctl start mysqld.service
[root@zgg script]# /bin/systemctl status mysqld.service
● mysqld.service - MySQL Server
   Loaded: loaded (/usr/lib/systemd/system/mysqld.service; enabled; vendor preset: disabled)
   Active: active (running) since 五 2020-10-16 23:32:51 CST; 1min 58s ago
     Docs: man:mysqld(8)
           http://dev.mysql.com/doc/refman/en/using-systemd.html
  Process: 6930 ExecStartPre=/usr/bin/mysqld_pre_systemd (code=exited, status=0/SUCCESS)
 Main PID: 7005 (mysqld)
   Status: "Server is operational"
   CGroup: /system.slice/mysqld.service
           └─7005 /usr/sbin/mysqld

10月 16 23:32:47 zgg systemd[1]: Starting MySQL Server...
10月 16 23:32:51 zgg systemd[1]: Started MySQL Server.
```

---------------------------------------------------------------------------------

> 启动mysql8.0，报错，出现`ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: YES)`

解决方法：

1、修改MySQL 登入限制

```sh
# 在/etc/my.cnf文件下的[mysqld]的末尾追加上一句：skip-grant-tables 
[root@zgg ~]# cat /etc/my.cnf
# For advice on how to change settings please see
# http://dev.mysql.com/doc/refman/8.0/en/server-configuration-defaults.html

[mysql]
default-character-set=utf8

[mysqld]
skip-grant-tables 
....
```

2、重新启动MySQL服务

```sh
[root@zgg script]# /bin/systemctl restart mysqld.service      
```

3、登入MySQL，修改密码设置规则，修改密码

```sh
[root@zgg script]# mysql
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 7
Server version: 8.0.21 MySQL Community Server - GPL

Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> flush privileges;
Query OK, 0 rows affected (0.01 sec)

mysql> select host,user,plugin,authentication_string from mysql.user;
+-----------+------------------+-----------------------+------------------------------------------------------------------------+
| host      | user             | plugin                | authentication_string                                                  |
+-----------+------------------+-----------------------+------------------------------------------------------------------------+
| localhost | mysql.infoschema | caching_sha2_password | $A$005$THISISACOMBINATIONOFINVALIDSALTANDPASSWORDTHATMUSTNEVERBRBEUSED |
| localhost | mysql.session    | caching_sha2_password | $A$005$THISISACOMBINATIONOFINVALIDSALTANDPASSWORDTHATMUSTNEVERBRBEUSED |
| localhost | mysql.sys        | caching_sha2_password | $A$005$THISISACOMBINATIONOFINVALIDSALTANDPASSWORDTHATMUSTNEVERBRBEUSED |
| localhost | root             | caching_sha2_password | $A$005$
                                                                1(z=SER s1(BQmzoDYXKEZzrJWGMJUVajerO272z57C3YQGhvN/5ou. |
+-----------+------------------+-----------------------+------------------------------------------------------------------------+
4 rows in set (0.00 sec)

mysql> ALTER user 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY '123456';
ERROR 1819 (HY000): Your password does not satisfy the current policy requirements
mysql> SHOW VARIABLES LIKE 'validate_password%'; 
+--------------------------------------+--------+
| Variable_name                        | Value  |
+--------------------------------------+--------+
| validate_password.check_user_name    | ON     |
| validate_password.dictionary_file    |        |
| validate_password.length             | 8      |
| validate_password.mixed_case_count   | 1      |
| validate_password.number_count       | 1      |
| validate_password.policy             | MEDIUM |
| validate_password.special_char_count | 1      |
+--------------------------------------+--------+
7 rows in set (0.01 sec)

mysql> set global validate_password.policy=0;
Query OK, 0 rows affected (0.00 sec)

mysql> set global validate_password.length=4;
Query OK, 0 rows affected (0.00 sec)

mysql> flush privileges;
Query OK, 0 rows affected (0.00 sec)

mysql> ALTER user 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY '1234';
Query OK, 0 rows affected (0.01 sec)

mysql> flush privileges;
Query OK, 0 rows affected (0.00 sec)
```

4、再次修改MySQL 登入限制

去除[mysqld] 中的skip-grant-tables ，重新启动MySQL服务。

```sh
[root@zgg script]# cat mysql_start.py 
import os

# python mysql_start.py
# password:1234
start = "/bin/systemctl start mysqld.service && mysql -uroot -p1234"

os.system(start)
   
[root@zgg script]# python mysql_start.py 
mysql: [Warning] Using a password on the command line interface can be insecure.
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 8
Server version: 8.0.21 MySQL Community Server - GPL

Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
| sys                |
+--------------------+
4 rows in set (0.00 sec)

mysql> 
```

参考：[1](https://blog.csdn.net/zhouzhiwengang/article/details/87378046?utm_medium=distribute.pc_relevant.none-task-blog-BlogCommendFromMachineLearnPai2-2.channel_param&depth_1-utm_source=distribute.pc_relevant.none-task-blog-BlogCommendFromMachineLearnPai2-2.channel_param) [2](https://blog.csdn.net/weixin_42955916/article/details/104670182?utm_medium=distribute.pc_relevant_t0.none-task-blog-BlogCommendFromMachineLearnPai2-1.channel_param&depth_1-utm_source=distribute.pc_relevant_t0.none-task-blog-BlogCommendFromMachineLearnPai2-1.channel_param)

---------------------------------------------------------------------------------

> 当 `sc = SparkContext(conf)`，会报如下错误:
`org.apache.spark.SparkException: Could not parse Master URL: '<pyspark.conf.SparkConf object at 0x106666390>'`

解决：阅读代码，发现它的构造函数声明如下所示：

    def __init__(self, master=None, appName=None, sparkHome=None, pyFiles=None,
        environment=None, batchSize=0, serializer=PickleSerializer(), conf=None,
        gateway=None, jsc=None, profiler_cls=BasicProfiler):

而前面的代码仅仅是简单的将 conf 传递给 SparkContext 构造函数，这就会导致 Spark 会将 conf 看做是 master 参数的值，即默认为第一个参数。所以这里要带名参数：`sc = SparkContext(conf = conf)`

---------------------------------------------------------------------------------

> 当打包 hadoop wordcount 后，在集群执行，出现 `java.lang.SecurityException: Invalid signature file digest for Manifest main attributes` 错误。

原因：因为依赖 jar 包中的 META-INF 中有多余的 .SF 文件与当前 jar 包冲突。

法1：在打包前删除依赖jar包的.SF文件

法2：在打完的jar包执行 `zip -d your.jar 'META-INF/.SF' 'META-INF/.RSA' 'META-INF/*SF'`

--------------------------------------------------------------------------

> 在使用sqoop2，从mysql传递数据到hdfs中时:

(1)出现`Caused by: Exception: java.sql.SQLException Message: Access denied for user 'root'@'zgg' (using password: YES`

需要创建用户，并赋予权限

```sql
mysql> CREATE USER 'root'@'zgg' IDENTIFIED BY '1234';               
Query OK, 0 rows affected (0.02 sec)

mysql> grant all privileges on *.* to 'root'@'zgg';
Query OK, 0 rows affected (0.03 sec)

mysql> flush privileges;
Query OK, 0 rows affected (0.01 sec)
```

(2)出现`User: root is not allowed to impersonate root`错误。

在`core-site.xml`里，将"xxx"替换成root即可，再重启hdfs

```xml
<property>
        <name>hadoop.proxyuser.xxx.hosts</name>
        <value>*</value>
    </property>
    <property>
        <name>hadoop.proxyuser.xxx.groups</name>
        <value>*</value>
    </property>
```

`*`表示可通过超级代理"xxx"操作hadoop的用户、用户组和主机

(3)出现`ClassNotFoundException: org.apache.commons.lang.StringUtils`

在lib目录下，分别测试放了`commons-lang-2.6.jar`和`commons-lang3-3.11.jar`。还是出现问题。

【未解决】

----------------------------------------------------------------------

> sqoop版本更换为1.4.7时:

(1)出现了`ClassNotFoundException: org.apache.commons.lang.StringUtils`问题，把lib目录下的`commons-lang3-3.4.jar `换成了`commons-lang-2.6.jar`，问题解决。


(2)出现`ClassNotFoundException: org.apache.hadoop.hive.conf.HiveConf`，只需`cp hive-common-3.1.2.jar /opt/sqoop-1.4.7.bin__hadoop-2.6.0/lib`

---------------------------------------------------------------------

> 启动hbase，出现`java.net.BindException: 地址已在使用`

hbase 会自带 zookeeper，报错的原因是 zookeeper 启动冲突了，hbase 启动并一起启动 zookeeper，这时应该是 zookeeper 已经启动了，需要停止 zookeeper。

在 hbase-env.sh 里面设置 HBASE_MANAGES_ZK 改成 false，默认是 true，先手动启动 zookeeper，然后再启动 hbase。

---------------------------------------------------------------------

> 启动hbase，HMaster闪退，查看日志出现`java.lang.IllegalStateException: The procedure WAL relies on the ability to hsync for proper operation during component failures, but the underlying filesystem does not support doing so. Please check the config value of 'hbase.procedure.store.wal.use.hsync' to set the desired level of robustness and ensure the config value of 'hbase.wal.dir' points to a FileSystem mount that can provide it.`

设为 false

```xml
<property>
  <name>hbase.unsafe.stream.capability.enforce</name>
  <value>false</value>
</property>
```

---------------------------------------------------------------------

flume 采集数据到 hdfs：

问题1：文件内容出现乱码。

解决：添加属性，以流方式写入。

    a1.sinks.k1.hdfs.fileType=DataStream

问题2：出现`java.lang.NullPointerException: Expected timestamp in the Flume event headers, but it was null`

解决：添加如下内容

    a1.sinks.k1.hdfs.useLocalTimeStamp = true

---------------------------------------------------------------------

flume 采集数据到 hive：

问题：出现`MetaException(message:Could not connect to meta store using any of the URIs provided. Most recent fa...`

解决：在 hive-site.xml 中配置 hive.metastore.uris 属性

```xml
<property>
    <name>hive.metastore.uris</name>
    <value>thrift://zgg:9083</value>
</property>

```

---------------------------------------------------------------------

使用 zookeeper 集群启动脚本启动集群，一直失败，出现`Error contacting service. It is probably not running`

因为脚本在执行ssh时，不会加载被ssh的机器上的系统环境变量 source  /etc/profile

但是它会加载用户的环境变量 `source  ~/.bashrc`  (/home/admin/.bashrc)

所以我们可以把系统中的环境变量追加到用户的环境变量中去：

    cat /etc/profile >> ~/.bashrc

原文地址：[https://blog.csdn.net/stupid_qiu/article/details/86496173](https://blog.csdn.net/stupid_qiu/article/details/86496173)

---------------------------------------------------------------------

安装完mysql，执行 `mysql -u root -p` 后，出现 `mysql: [ERROR] unknown variable 'datadir=/var/lib/mysql'` 错误。

错误跟 `my.cnf` 文件有关，需要这样：

    [mysql]
    default-character-set=utf8

    [mysqld]

而不是：

    [mysqld]

    [mysql]
    default-character-set=utf8

---------------------------------------------------------------------

在 shell 脚本里设置启动 alluxio 集群的命令，执行后显示 `./alluxio.sh:行5: alluxio-start.sh all: 未找到命令`

安装完 hive-2.3.5 后，执行 `bin/schematool -dbType mysql -initSchema`，出现 `Exception in thread "main" java.lang.ClassNotFoundException: org.apache.hive.beeline.HiveSchemaTool`

---------------------------------------------------------------------

使用 `spark-submit` 执行程序时，出现 `ExitCodeException exitCode=13`，错误原因是搭建的 Spark 集群模式和在执行命令中指定的集群模式不一致。

---------------------------------------------------------------------

在 django 项目中，使用服务器 IP 地址打开，出现 `Invalid HTTP_HOST header: 'xxx.xx.xxx.xxx:8000'. You may need to add 'xxx.xx' to ALLOWED_HOSTS！`，需要修改创建项目时生成的 setting.py 文件，将 `ALLOWED_HOSTS = []` 改为 `ALLOWED_HOSTS = ['*']` ，再次运行即可成功访问。

---------------------------------------------------------------------

在 servlet 的登录系统案例中，执行拼接的语句出现 `java.sql.SQLSyntaxErrorException: This version of MySQL doesn't yet support` 错误。

再加一层子查询

    ps = conn.prepareStatement("select * from users " +
                        "where userId in (" +
                            "select userId from (" +
                                "select userId from users " +
                                    "limit "+pageSize*(pageNow-1)+","+pageSize+") t) ");


---------------------------------------------------------------------

使用 cx_Oracle 库远程连接 ORACLE，出现 `Cannot locate a 64-bit Oracle Client library` 错误。

解决：除了安装、配置 instantclient 外，还需要把 instantclient 目录下的 oci.dll、oraocci11.dll、oraociei11.dll 放到 python 的 Lib 下的 site-packages 下。同时注意版本一致。